File: internal/config/state_paths_test.go
package config

import (
	"errors"
	"path/filepath"
	"strings"
	"testing"
)

func TestDefaultStateLayoutUsesSingleLocalEnvironment(t *testing.T) {
	restoreHome := stubUserHomeDir("/Users/tester", nil)
	restoreEnv := stubLookupEnv("", false)
	defer restoreHome()
	defer restoreEnv()

	layout, err := DefaultStateLayout()
	if err != nil {
		t.Fatalf("DefaultStateLayout() error = %v", err)
	}

	wantStateRoot := filepath.Clean("/Users/tester/.redeven")
	wantStateDir := filepath.Join(wantStateRoot, "local-environment")
	if layout.StateRoot != wantStateRoot {
		t.Fatalf("StateRoot = %q", layout.StateRoot)
	}
	if layout.ConfigPath != filepath.Join(wantStateDir, "config.json") {
		t.Fatalf("ConfigPath = %q", layout.ConfigPath)
	}
	if layout.SecretsPath != filepath.Join(wantStateDir, "secrets.json") {
		t.Fatalf("SecretsPath = %q", layout.SecretsPath)
	}
	if layout.LockPath != filepath.Join(wantStateDir, "agent.lock") {
		t.Fatalf("LockPath = %q", layout.LockPath)
	}
	if layout.StateDir != wantStateDir {
		t.Fatalf("StateDir = %q", layout.StateDir)
	}
	if layout.RuntimeControlSocketPath != filepath.Join(wantStateDir, "runtime", "control.sock") {
		t.Fatalf("RuntimeControlSocketPath = %q", layout.RuntimeControlSocketPath)
	}
	if layout.RuntimeMaintenancePath != filepath.Join(wantStateDir, "runtime", "maintenance", "current.json") {
		t.Fatalf("RuntimeMaintenancePath = %q", layout.RuntimeMaintenancePath)
	}
	if RuntimeMaintenancePathFromConfigPath(layout.ConfigPath) != layout.RuntimeMaintenancePath {
		t.Fatalf("RuntimeMaintenancePathFromConfigPath() = %q, want %q", RuntimeMaintenancePathFromConfigPath(layout.ConfigPath), layout.RuntimeMaintenancePath)
	}
	if layout.DiagnosticsDir != filepath.Join(wantStateDir, "diagnostics") {
		t.Fatalf("DiagnosticsDir = %q", layout.DiagnosticsDir)
	}
	if layout.AuditDir != filepath.Join(wantStateDir, "audit") {
		t.Fatalf("AuditDir = %q", layout.AuditDir)
	}
	if layout.AppsDir != filepath.Join(wantStateDir, "apps") {
		t.Fatalf("AppsDir = %q", layout.AppsDir)
	}
}

func TestResolveStateRootUsesEnvOverride(t *testing.T) {
	restoreHome := stubUserHomeDir("/Users/ignored", nil)
	restoreEnv := stubLookupEnv("/tmp/redeven-state", true)
	defer restoreHome()
	defer restoreEnv()

	root, err := ResolveStateRoot("")
	if err != nil {
		t.Fatalf("ResolveStateRoot() error = %v", err)
	}
	if root != filepath.Clean("/tmp/redeven-state") {
		t.Fatalf("ResolveStateRoot() = %q", root)
	}
}

func TestLocalEnvironmentStateLayoutUsesStateRootOverride(t *testing.T) {
	layout, err := LocalEnvironmentStateLayout("/tmp/redeven-profile")
	if err != nil {
		t.Fatalf("LocalEnvironmentStateLayout() error = %v", err)
	}

	wantStateRoot := filepath.Clean("/tmp/redeven-profile")
	wantStateDir := filepath.Join(wantStateRoot, "local-environment")
	if layout.StateRoot != wantStateRoot {
		t.Fatalf("StateRoot = %q", layout.StateRoot)
	}
	if layout.ConfigPath != filepath.Join(wantStateDir, "config.json") {
		t.Fatalf("ConfigPath = %q", layout.ConfigPath)
	}
}

func TestLocalEnvironmentStateLayoutShortensLongRuntimeControlSocketPath(t *testing.T) {
	stateRoot := filepath.Join(t.TempDir(), strings.Repeat("long-state-segment-", 8))
	layout, err := LocalEnvironmentStateLayout(stateRoot)
	if err != nil {
		t.Fatal(err)
	}

	legacyPath := filepath.Join(layout.StateDir, "runtime", "control.sock")
	if len([]byte(legacyPath)) <= maxUnixSocketPathBytes {
		t.Fatalf("test fixture path length = %d, want > %d", len([]byte(legacyPath)), maxUnixSocketPathBytes)
	}
	if layout.RuntimeControlSocketPath == legacyPath {
		t.Fatalf("RuntimeControlSocketPath retained overlong path %q", layout.RuntimeControlSocketPath)
	}
	if len([]byte(layout.RuntimeControlSocketPath)) > maxUnixSocketPathBytes {
		t.Fatalf("RuntimeControlSocketPath length = %d, want <= %d: %q", len([]byte(layout.RuntimeControlSocketPath)), maxUnixSocketPathBytes, layout.RuntimeControlSocketPath)
	}
	if filepath.Ext(layout.RuntimeControlSocketPath) != ".sock" {
		t.Fatalf("RuntimeControlSocketPath = %q, want .sock suffix", layout.RuntimeControlSocketPath)
	}
	if got := RuntimeControlSocketPathFromConfigPath(layout.ConfigPath); got != layout.RuntimeControlSocketPath {
		t.Fatalf("RuntimeControlSocketPathFromConfigPath() = %q, want %q", got, layout.RuntimeControlSocketPath)
	}

	second, err := LocalEnvironmentStateLayout(stateRoot)
	if err != nil {
		t.Fatal(err)
	}
	if second.RuntimeControlSocketPath != layout.RuntimeControlSocketPath {
		t.Fatalf("short Runtime control socket is not deterministic: first=%q second=%q", layout.RuntimeControlSocketPath, second.RuntimeControlSocketPath)
	}
	if filepath.Dir(layout.RuntimeControlSocketPath) != "/tmp" {
		t.Fatalf("RuntimeControlSocketPath directory = %q, want /tmp", filepath.Dir(layout.RuntimeControlSocketPath))
	}
}

func TestLongRuntimeControlSocketPathDoesNotDependOnProcessTempEnvironment(t *testing.T) {
	stateRoot := filepath.Join(t.TempDir(), strings.Repeat("cross-process-state-segment-", 8))
	t.Setenv("TMPDIR", filepath.Join(t.TempDir(), "gateway-temp"))
	gatewayLayout, err := LocalEnvironmentStateLayout(stateRoot)
	if err != nil {
		t.Fatal(err)
	}
	t.Setenv("TMPDIR", filepath.Join(t.TempDir(), "desktop-temp"))
	desktopLayout, err := LocalEnvironmentStateLayout(stateRoot)
	if err != nil {
		t.Fatal(err)
	}

	if gatewayLayout.RuntimeControlSocketPath != desktopLayout.RuntimeControlSocketPath {
		t.Fatalf(
			"Runtime control socket changed across process temp environments: gateway=%q desktop=%q",
			gatewayLayout.RuntimeControlSocketPath,
			desktopLayout.RuntimeControlSocketPath,
		)
	}
}

func TestDefaultStateLayoutReturnsMissingHomeError(t *testing.T) {
	restoreHome := stubUserHomeDir("", errors.New("home missing"))
	restoreEnv := stubLookupEnv("", false)
	defer restoreHome()
	defer restoreEnv()

	_, err := DefaultStateLayout()
	if !errors.Is(err, ErrHomeDirUnavailable) {
		t.Fatalf("DefaultStateLayout() error = %v, want ErrHomeDirUnavailable", err)
	}
}

func stubUserHomeDir(home string, err error) func() {
	previous := userHomeDir
	userHomeDir = func() (string, error) {
		return home, err
	}
	return func() {
		userHomeDir = previous
	}
}

func stubLookupEnv(value string, ok bool) func() {
	previous := lookupEnv
	lookupEnv = func(key string) (string, bool) {
		if key == stateRootEnvName {
			return value, ok
		}
		return "", false
	}
	return func() {
		lookupEnv = previous
	}
}

File: desktop/src/main/statePaths.test.ts
import path from 'node:path';

import { describe, expect, it } from 'vitest';

import {
  defaultLocalEnvironmentStateLayout,
  localEnvironmentStateLayout,
  resolveConfiguredDesktopCacheRoot,
  resolveConfiguredDesktopTempRoot,
  resolveConfiguredDesktopUserDataRoot,
  runtimeControlSocketPath,
} from './statePaths';

describe('statePaths', () => {
  it('resolves the default Local Environment state layout under the single layout', () => {
    expect(defaultLocalEnvironmentStateLayout({ HOME: '/Users/tester' }, () => '/ignored')).toEqual({
      stateRoot: '/Users/tester/.redeven',
      configPath: '/Users/tester/.redeven/local-environment/config.json',
      secretsFile: '/Users/tester/.redeven/local-environment/secrets.json',
      lockFile: '/Users/tester/.redeven/local-environment/agent.lock',
      stateDir: '/Users/tester/.redeven/local-environment',
      runtimeControlSocket: '/Users/tester/.redeven/local-environment/runtime/control.sock',
      diagnosticsDir: '/Users/tester/.redeven/local-environment/diagnostics',
      auditDir: '/Users/tester/.redeven/local-environment/audit',
      appsDir: '/Users/tester/.redeven/local-environment/apps',
      gatewayDir: '/Users/tester/.redeven/local-environment/gateway',
    });
  });

  it('resolves the explicit Local Environment layout helper to the same single layout', () => {
    expect(localEnvironmentStateLayout({ HOME: '/Users/tester' }, () => '/ignored')).toEqual(expect.objectContaining({
      configPath: '/Users/tester/.redeven/local-environment/config.json',
      stateDir: '/Users/tester/.redeven/local-environment',
      runtimeControlSocket: '/Users/tester/.redeven/local-environment/runtime/control.sock',
    }));
  });

  it('fails clearly when no home directory is available for implicit defaults', () => {
    expect(() => defaultLocalEnvironmentStateLayout({}, () => '')).toThrow('user home directory is unavailable');
  });

  it('gives an explicit development state root priority over the user profile fallback', () => {
    expect(defaultLocalEnvironmentStateLayout({
      HOME: '/Users/tester',
      REDEVEN_STATE_ROOT: '/tmp/redeven-dev-checkout',
    }, () => '/ignored')).toEqual(expect.objectContaining({
      stateRoot: '/tmp/redeven-dev-checkout',
      stateDir: '/tmp/redeven-dev-checkout/local-environment',
    }));
  });

  it('resolves an explicit task-owned Desktop temp root without falling back to the system temp directory', () => {
    expect(resolveConfiguredDesktopTempRoot({
      REDEVEN_DESKTOP_TEMP_ROOT: '/tmp/redeven-plugin-e2e/runtime-data',
      TMPDIR: '/var/folders/shared',
    })).toBe('/tmp/redeven-plugin-e2e/runtime-data');
    expect(resolveConfiguredDesktopTempRoot({ TMPDIR: '/var/folders/shared' })).toBeNull();
  });

  it('resolves task-owned Electron user-data and cache roots without falling back to the shared profile', () => {
    const env = {
      REDEVEN_DESKTOP_USER_DATA_ROOT: '/tmp/redeven-scope/user-data',
      REDEVEN_DESKTOP_CACHE_ROOT: '/tmp/redeven-scope/cache',
    };
    expect(resolveConfiguredDesktopUserDataRoot(env)).toBe('/tmp/redeven-scope/user-data');
    expect(resolveConfiguredDesktopCacheRoot(env)).toBe('/tmp/redeven-scope/cache');
    expect(resolveConfiguredDesktopUserDataRoot({})).toBeNull();
    expect(resolveConfiguredDesktopCacheRoot({})).toBeNull();
  });

  it.each(['darwin', 'linux'] as const)('uses a stable short control socket for long %s state roots', (platform) => {
    const stateRoot = path.join('/private/tmp', 'redeven-state-segment-'.repeat(8));
    const stateDir = path.join(stateRoot, 'local-environment');
    const socketPath = runtimeControlSocketPath(stateDir, platform);

    expect(Buffer.byteLength(path.join(stateDir, 'runtime', 'control.sock'))).toBeGreaterThan(100);
    expect(Buffer.byteLength(socketPath)).toBeLessThanOrEqual(100);
    expect(socketPath).toMatch(/^\/tmp\/redeven-runtime-[a-f0-9]{24}\.sock$/);
    expect(runtimeControlSocketPath(stateDir, platform)).toBe(socketPath);
  });

  it('does not expose a second local Flower state layout', async () => {
    const paths = await import('./statePaths');
    expect(Object.keys(paths).filter((key) => key.toLowerCase().includes('flower'))).toEqual([]);
  });
});
